### DISCUSSION TITLE
分流解析疑问

### DISCUSSION THREAD

**Discussion by hussion** [kkkgo/PaoPaoDNS#70] [问答和帮助] [answered]
看了一下paopaodns在分流场景下，mosdns的配置逻辑：

`启动mosdns进行分流处理（不缓存任何请求，缓存由redis完成），处理过程如下：
拒绝qtype65请求（避免apple设备类型被动使用https DNS）
读取force_nocn_list.txt的域名列表，直接转发到unbound-dnscrypt。
读取force_cn_list.txt的域名列表，直接转发到unbound-local。
先尝试请求unbound-local，如果返回的结果IP位于China大陆，则接受并返回到客户端。
如果返回的结果不位于China大陆，则把结果抛弃并转发到unbound-dnscrypt。
  #@如果请求的域名使用了unbound-dnscrypt，并且同时具备IPv4和IPv6的解析，那么将会仅返回IPv4记录，过滤IPv6记录，如果仅有IPv6记录，将不会过滤，这个流程是考虑双栈用户的分流问题，因为应用端会优先使用IPv6地址而导致你的静态路由失效，但又不放弃那些仅有IPv6的网站。`

默认只加载了一份CN的ipset，为什么不先加载一份CN的domainset，先进行域名匹配，匹配不到再进行DNS解析，然后进行IP比对是否属于CN 呢?
这样对于命中的CN Site 不是少了一层不必要的DNS解析吗？

另外，对于部分PT网站，通过`force_cn_list.txt`指定强制本地解析后，会出现解析失败的问题，是什么原因呢？

**Comment by kkkgo** [comment, marked as answer]
>1、默认只加载了一份CN的ipset，为什么不先加载一份CN的domainset，先进行域名匹配，匹配不到再进行DNS解析，然后进行IP比对是否属于CN 呢?

- 你说的是`USE_MARK_DATA`功能，实现方式跟你说的有点区别，但效率和精准度更高。
- 另外IP比对是基于redis缓存，实际上不需要等待DNS解析完成后再比对（仅首次查询该域名需要）。

>2.另外，对于部分PT网站，通过`force_cn_list.txt`指定强制本地解析后，会出现解析失败的问题，是什么原因呢？

因为你加入的pt网站的NS的IP地址实际上都不在大陆，你所在的网络环境无法直接连接权威服务器就会失败，这种场景你应该加入`force_nocn_list.txt`。如果你没有用到`AUTO_FORWARD`功能的话，他默认就是当`force_nocn_list.txt`处理，不需要添加。实际上，网站本身的IP也不在大陆，不应该加入`force_cn_list.txt`。如果是tracker，你可以加入`trackerslist.txt`。

**Reply by hussion** [reply]
我目前是搭配`paopaoGateway`一起来使用的，而且`USE_MARK_DATA`和`AUTO_FORWARD `均已打开，另外昨天初步试了一下， 发现iOS手机端打开淘宝APP，`taobao.com`以及`alicdn.com`等对应的三级域名，都被当做解析成了外站，流入了paopaogateway中。

1. 这种情况是`USE_MARK_DATA`对应的数据库没下载成功吗还是其他原因呢？
2. 如何检查`USE_MARK_DATA`是否生效 ？ 以及对应的`USE_MARK_DATA`数据库和geoip数据库是否下载成功？目前docker对应的文件夹里有看到`Country-only-cn-private.mmdb -- 242KB左右`和`global_mark.dat -- 1.1MB左右`
3. 请教一下`拒绝qtype65请求（避免apple设备类型被动使用https DNS）`这个如何理解？目前观测apple到相关的域名都走到`paopaoGateway`里了

**Reply by kkkgo** [reply]
USE_MARK_DATA理论上不会让taobao.com标记为非CN域名。你可以使用ADDINFO功能查看结果来源。可能是你配置的路由或者PPGW有问题。

**Reply by hussion** [reply]
另外再确认下，trackerslist.txt里添加域名的规则，和下面`force_cn_list.txt`的规则是不是一致的？我看现在生成的trackerslist.txt文件里都是`带协议+端口+path`的完整URL路径

```
force_cn_list.txt：强制使用本地递归服务器查询的域名列表，容器版本更新不会覆盖该文件。一行一条，语法规则如下：
以domain:开头域匹配: domain:03k.org会匹配自身03k.org，以及其子域名www.03k.org, blog.03k.org等。
以full:开头，完整匹配，full:03k.org 只会匹配自身。完整匹配优先级更高。
以regxp:开头，正则匹配，如regexp:.+\.03k\.org$。[Go标准正则](https://github.com/google/re2/wiki/Syntax)。
以keyword:开头匹配域名关键字，如以keyword: 03k.org会匹配到www.03k.org.cn
```

**Reply by kkkgo** [reply]
`trackerslist.txt`和`force_nocn_list.txt`的作用实际上是一样的。只不过`trackerslist.txt`会从url中提取域名再处理，并且接受数据更新。`trackerslist.txt`也可以共享给其他软件使用。

**Reply by hussion** [reply]
今天又重新试了下，把pt域名放到trackerlist.txt里了，但是依然会走到PPGW里，debug日志如下：
部署方式：群晖docker部署，网络为默认bridge方式

**Reply by kkkgo** [reply]
如果你是指pt网站的话，你可以放入`force_nocn_list`。trackerslist仅接受tracker格式，其他内容会过滤。
你可以dig 具体的pt域名来看解析结果。

**Reply by hussion** [reply]
把域名放到force_nocn_list里，依然不行，接着把cn_tracker关掉后正常了，初步猜测是系统生成的trackerlist解析可能有问题

**Reply by kkkgo** [reply]
cn_tracker不会影响你的force_nocn_list。你应该只是重启容器后生效了。

**Reply by hussion** [reply]
再请教一个问题: 我是群晖docker部署，网络为默认bridge方式，群晖本身的ip为192.168.50.3，为了让群晖都能科学上网，我把群晖本身dns改为192.168.50.3，群晖本身上网没问题，但是群晖上面运行的docker容器都无法联网了，也无法ping通192.168.50.3，这是什么原因呢？以及正确的设置方式应该是怎样的？

**Reply by kkkgo** [reply]
威联通有个qnet模式是直接分配到网卡里比较好用，群晖好像没这种东西，可以参考一下网上的一些群晖的macvlan教程。
https://post.smzdm.com/p/awkv40k4/
